src/Processors/QueryPlan/Optimizations/reduceAggregationKeysSize.cpp

```diff
@@ -0,0 +1,176 @@
+#include <Processors/QueryPlan/AggregatingStep.h>
+#include <Processors/QueryPlan/ExpressionStep.h>
+#include <Processors/QueryPlan/RollupStep.h>
+#include <Processors/QueryPlan/TotalsHavingStep.h>
+#include <Processors/QueryPlan/CubeStep.h>
+#include <Processors/QueryPlan/Optimizations/Optimizations.h>
+#include <Functions/FunctionFactory.h>
+#include <DataTypes/DataTypesNumber.h>
+
+namespace DB::QueryPlanOptimizations
+{
+
+namespace
+{
+
+std::shared_ptr<ActionsDAG> buildPredecessorActionsDag(const Block & header, const Names & keys, const DataHints & hints,
+        Names & changed_aggregating_keys, Names & new_aggregating_keys, DataTypes & changed_data_types)
+{
+    auto dag = std::make_shared<ActionsDAG>(header.getColumnsWithTypeAndName());
+    for (const auto & old_key : keys)
+    {
+        if (hints.contains(old_key) && hints.at(old_key).isRangeLengthLessOrEqualThan(65535))
+        {
+            const ActionsDAG::Node * key_column_node = dag->tryFindInOutputs(old_key);
+            const auto & key_column_type = key_column_node->result_type;
+            const auto & type_name = key_column_type->getName();
+
+            if (type_name == "UInt8" || type_name == "Int8" || type_name == "UInt16" || type_name == "Int16")
+            {
+                new_aggregating_keys.push_back(old_key);
```

Comment:
looking at l.155-156 it seems that `new_aggregating_keys` and `changed_aggregating_keys` are supposed to be of the same size. but it will not be true if we got into this branch, right?